python/tank/context.py

```diff
@@ -628,7 +628,102 @@ def create_copy_for_user(self, user):
+
+    ################################################################################################
+    # serialization
+
+    def serialize(self, with_user_credentials=True):
+        """
+        Serializes the context into a string.
+
+        Any Context object can be serialized to/deserialized from a string.
+        This can be useful if you need to pass a Context between different processes.
+        As an example, the ``tk-multi-launchapp`` uses this mechanism to pass the Context
+        from the launch process (e.g. for example Shotgun Desktop) to the
+        Application (e.g. Maya) being launched. Example:
+
+            >>> import sgtk
+            >>> tk = sgtk.sgtk_from_path("/studio.08/demo_project")
+            >>> ctx = tk.context_from_path("/studio.08/demo_project/sequences/AAA/ABC/Lighting/dirk.gently/work")
+            >>> context_str = ctx.serialize(ctx)
+            >>> new_ctx = sgtk.Context.deserialize(context_str)
+
+        :param with_user_credentials: If ``True``, the currently authenticated user's credentials, as
+            returned by :meth:`sgtk.get_authenticated_user`, will also be serialized with the context.
+
+        .. note:: For example, credentials should be omitted (``with_user_credentials=False``) when
+            serializing the context from a user's current session to send it to a render farm. By doing
+            so, invoking :meth:`sgtk.Context.deserialize` on the render farm will only restore the
+            context and not the authenticated user.
+
+        :returns: String representation
+        """
+        # Avoids cyclic imports
+        from .api import get_authenticated_user
+
+        data = {
+            "project": self.project,
+            "entity": self.entity,
+            "user": self.user,
+            "step": self.step,
+            "task": self.task,
+            "additional_entities": self.additional_entities,
+            "_pc_path": self.tank.pipeline_configuration.get_path()
+        }
+
+        if with_user_credentials:
+            # If there is an authenticated user.
+            user = get_authenticated_user()
+            if user:
+                # We should serialize it as well so that the next process knows who to
+                # run as.
+                data["_current_user"] = authentication.serialize_user(user)
+        return pickle.dumps(data)
+
+    @classmethod
+    def deserialize(cls, context_str):
+        """
+        The inverse of :meth:`Context.serialize`.
+
+        :param context_str: String representation of context, created with :meth:`Context.serialize`
+
+        .. note:: If the context was serialized with the user credentials, the currently authenticated
+            user will be updated with these credentials.
+
+        :returns: :class:`Context`
+        """
+        # lazy load this to avoid cyclic dependencies
+        from .api import Tank, set_authenticated_user
+
+        data = pickle.loads(context_str)
```

Comment:
minor: maybe we should put this in a try/catch and raise a more tk-specific exception (e.g 'could not deserialize toolkit context') rather than having the pickle error bubble up?